exo/inference/inference_engine.py

```diff
@@ -24,6 +24,22 @@ async def decode(self, shard: Shard, tokens: np.ndarray) -> str:
   async def infer_tensor(self, request_id: str, shard: Shard, input_data: np.ndarray) -> np.ndarray:
     pass
   
+  async def save_session(self, key, value):
+    self.session[key] = value
+  
+  async def ensure_session(self, key, check, value_gen, hook=None):
+    if key not in self.session or not check(self.session[key]):
+      await self.save_session(key, value_gen())
+      if hook is not None:
+        hook()
+  
+  async def ensure_session_match(self, key, check, value_gen):
+    if key not in self.session or not check(self.session[key]):
+      await self.save_session(key, value_gen())
+  
+  async def clear_session(self):
+    session.empty()
```

Comment:
should be `self.session`?

Reply:
Yea, though this actually didn't end up being used so I'm gonna get rid of it for simplification's sake